Add lint with `detekt` and `ktlint`, `twitter.compose` rules

alesrazym opened the pull request.

Commit:
FEAT: Update `detekt.yml` rules

Update rules to match `.editorconfig`
Ignore `@Composable` method naming (twitter's `TwitterCompose.ComposableNaming` applies.
Do not run libraries rules - turn off.
Force trailing comma.

Commit:
FIX: Code style & lint fixes by detekt

havlisimo commented on app/src/main/java/cz/quanti/razym/rocketapp/ui/theme/Color.kt:
v tomto pripade by bylo lepsi pouzit suppress na cely file
@file:Suppress("MagicNumber")

alesrazym replied:
jo, to dává smysl. Obecně bych měl trochu obavu, aby to pak nevynechalo místa, která zkontrolovat chceš. Úplně nejlepší by bylo, kdyby to jako magic number nehlásilo vůbec.

alesrazym replied:
vyřeším přidáním názvu parametru

havlisimo commented on app/src/main/java/cz/quanti/razym/rocketapp/ui/theme/Theme.kt:
```diff
@@ -34,7 +34,7 @@ private val LightColorScheme = lightColorScheme(
     onBackground = Color(0xFF1C1B1F),
     onSurface = Color(0xFF1C1B1F),
-    */
+     */
```
co je toto za konvenci 😃

alesrazym replied:
hvězdička pod hvězdičkou..

havlisimo replied:
ajo, ok

havlisimo commented on app/build.gradle.kts:
```diff
@@ -2,6 +2,16 @@ plugins {
+    toolVersion = "1.23.3"
+    buildUponDefaultConfig = true // preconfigure defaults
+    allRules = false // activate all available (even unstable) rules.
+    config.setFrom("$rootDir/detekt.yml") // point to your custom config defining rules to run, overwriting default behavior
```
v momente kdy pouzijes
buildUponDefaultConfig = true
neni potreba v tom detekt.yml definovat behavior pro vsechno, jak pises v komentari staci jen to co potrebujes overridovat

alesrazym replied:
jo to mi tady zbyly komentáře z guidy :( 

Byl bych pro následující:

- `detekt.yml` obsahuje všechna známá nastavení (vygenerováno pomocí gradle tasku)
- změna verze `toolVersion` tak nezmění chování linteru
- při přechodu na vyšší verzi nebo přidáním pravidel přegenerovat základní nastavení pomocí gradle tasku (je potřeba smazat/přejmenovat `detekt.yml` a zkontrolovat co se změnilo)

Nebo, jak to máte na projektech do teď?

havlisimo replied:
doted to na projektech nemame, ale tvoje vysvetleni dava smysl. Jak se potom resi kdyz si nejaka ta nastaveni upravime a pak prijde nova verze? 
- vygeneruje se novy soubor a manualne se to projde 
- nebo jde specifikovat nejaky detekt-quanti.yml, ktery by jeste overridoval ten detekt.yml?

alesrazym commented:
ad https://github.com/alesrazym/rocket-app/pull/8#discussion_r1400393688

> doted to na projektech nemame, ale tvoje vysvetleni dava smysl. Jak se potom resi kdyz si nejaka ta nastaveni upravime a pak prijde nova verze?
> 
>     * vygeneruje se novy soubor a manualne se to projde
> 
>     * nebo jde specifikovat nejaky detekt-quanti.yml, ktery by jeste overridoval ten detekt.yml?

Ano, řešil bych to manuálně, např. následujícím postupem:

1. Rename `detekt.yml` to e.g. `detekt_custom.yml`
2. Run `detektGenerateConfig`, rename output detekt.yml to `detect_old.yml`
3. Update ruleset, run `detektGenerateConfig`, rename output detekt.yml to `detect_new.yml`
4. Use diff to check (and apply) changes.
    - I.e. check old and new are the same
    - Or check all 3. In Android Studio, select them one by one with Ctrl/Cmd. Order matters. Then use context menu or Ctrl/Cmd + D to see 3way diff. Apply changes that make sense.
    - Finally,  drop new and old, rename `detekt_custom.yml` back to `detekt.yml`

Commit:
FEAT: Code style using `io.gitlab.arturbosch.detekt 1.23.4`.

Add `com.twitter.compose.rules:detekt:0.0.26` rules.
Add `com.autonomousapps.dependency-analysis` plugin to check dependencies health (e.g. unused dependencies).
Activate `detekt` plugin in Android Studio.
Add `detekt.yml` rules to IDEA detekt plugin configuration.

Commit:
WIP: Add detekt readme

Commit:
LIB: `detekt` & `ktlint` configuration update and readme

Add `detekt.md` readme.
Add Gradle plugins `kotlinter-gradle` and `ktlint-gradle` (not used now - slower).
Configure detekt baseline.

Commit:
format xml

Commit:
code style

Commit:
FIX: Code style after configuration update

Commit:
LIB: Migrate from Twitter's Jetpack Compose Rules to Jetpack Compose Rules

Commit:
FEAT: Add Gradle Convention Compose Plugin

Commit:
FEAT: Add Base Gradle Convention Plugin

Commit:
FEAT: Add Detekt and Kotlinter to Conventional Plugins, rfc

Add `pluginId()` extension method and use.

alesrazym commented:
@havlisimo Jak jsem již zmínil, nakonec jsem rozdělil `detekt` a `ktlint`, protože detekt se na formátovacích pravidlech zacyklil. Přidal jsem `detekt.md`, tak koukni, zda je to pochopitelné.

Dál jsem updatnul Jetpack Compose Rules a předělal konfiguraci na `Conventional Plugins`.

Prosím teda o review ještě jednou. Dík

Commit:
FIX: Add `ij_*` rules to `.editorconfig` to have the same code style on all local machines.

Note, that there still may be some rules missing, there were two rules found that have not been exported by IDE

Commit:
LIB: Update compose rules to `0.3.8`

alesrazym commented:
Branch rebased on compose branch, closing this PR and opening new one
